File: src/contributors.py
"""REST contributor fetch + bot strip + cache I/O.

Plan: parallel-readme-and-solo-filter Phase 1 (F2-F5).

Solo detection uses the REST endpoint
``GET /repos/{owner}/{repo}/contributors?per_page=2`` (cheaper than counting
the full list — len <= 1 after bot strip is enough). Results are cached on
disk keyed by ``repo_name + pushed_at``; stale entries (different pushed_at)
are ignored.
"""
from __future__ import annotations

import json
import re
import subprocess
from concurrent.futures import ThreadPoolExecutor
from pathlib import Path
from typing import Iterable

from src.selection import Repo

# ---------------------------------------------------------------------------
# Bot detection
# ---------------------------------------------------------------------------

_BOT_RE = re.compile(
    r"(.*\[bot\]$|^dependabot(-preview)?$|^github-actions$)",
    re.IGNORECASE,
)


def is_bot(login: str) -> bool:
    return bool(_BOT_RE.match(login))


def strip_bots(logins: Iterable[str]) -> tuple[str, ...]:
    return tuple(l for l in logins if not is_bot(l))


# ---------------------------------------------------------------------------
# Cache I/O
# ---------------------------------------------------------------------------

def cache_key(repo_name: str, pushed_at: str) -> str:
    return f"{repo_name}@{pushed_at}"


def load_cache(path: Path) -> dict[str, list[str]]:
    """Read JSON cache; return ``{}`` on missing or corrupt file."""
    try:
        text = path.read_text()
    except (FileNotFoundError, OSError):
        return {}
    try:
        data = json.loads(text)
    except (json.JSONDecodeError, ValueError):
        return {}
    if not isinstance(data, dict):
        return {}
    return {k: list(v) for k, v in data.items() if isinstance(v, list)}


def save_cache(path: Path, cache: dict[str, list[str]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(cache, indent=2, sort_keys=True))


# ---------------------------------------------------------------------------
# REST fetch
# ---------------------------------------------------------------------------

def fetch_contributors(owner: str, name: str) -> tuple[str, ...]:
    """Return tuple of human contributor logins (bots stripped).

    Empty tuple means: 0 contributors (empty repo / 404) OR 0 humans after
    bot strip — both count as solo per F3.
    """
    cmd = [
        "gh", "api",
        f"/repos/{owner}/{name}/contributors?per_page=2",
    ]
    try:
        result = subprocess.run(cmd, capture_output=True, check=True, timeout=15)
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired):
        # 404 (empty repo), 403, timeout, etc. — treat as 0 contributors.
        return ()
    try:
        payload = json.loads(result.stdout or b"[]")
    except (json.JSONDecodeError, ValueError):
        return ()
    if not isinstance(payload, list):
        return ()
    logins = [
        entry.get("login", "")
        for entry in payload
        if isinstance(entry, dict) and entry.get("login")
    ]
    return strip_bots(logins)


# ---------------------------------------------------------------------------
# Parallel enrichment
# ---------------------------------------------------------------------------

def enrich_repos(
    repos: list[Repo],
    *,
    owner: str,
    cache_path: Path,
    max_workers: int = 10,
) -> list[Repo]:
    """Return new Repo list with ``contributors`` populated.

    Cache hits skip the REST call. Misses are fetched in parallel
    (ThreadPoolExecutor, default 10 workers per F4). Cache is rewritten
    atomically once at the end.
    """
    cache = load_cache(cache_path)

    def resolve(repo: Repo) -> tuple[Repo, str, list[str]]:
        key = cache_key(repo.name, repo.pushed_at)
        cached = cache.get(key)
        if cached is not None:
            return repo, key, cached
        fetched = list(fetch_contributors(owner, repo.name))
        return repo, key, fetched

    enriched: list[Repo] = []
    new_entries: dict[str, list[str]] = {}

    with ThreadPoolExecutor(max_workers=max(1, max_workers)) as pool:
        for repo, key, logins in pool.map(resolve, repos):
            new_entries[key] = logins
            from dataclasses import replace
            enriched.append(replace(repo, contributors=tuple(logins)))

    # Persist (preserves cached entries we used + adds fresh ones).
    cache.update(new_entries)
    save_cache(cache_path, cache)
    return enriched

File: go/internal/contributors/contributors.go
// Package contributors fetches contributor lists with bot strip + on-disk cache.
package contributors

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sync"
)

var botRe = regexp.MustCompile(`(?i)(.*\[bot\]$|^dependabot(-preview)?$|^github-actions$)`)

// IsBot reports whether login matches the bot heuristic.
func IsBot(login string) bool { return botRe.MatchString(login) }

// StripBots returns logins minus any bot identities.
func StripBots(logins []string) []string {
	out := make([]string, 0, len(logins))
	for _, l := range logins {
		if !IsBot(l) {
			out = append(out, l)
		}
	}
	return out
}

// CacheKey returns "name@pushed_at" used by the on-disk cache.
func CacheKey(name, pushedAt string) string { return name + "@" + pushedAt }

// LoadCache reads JSON cache; returns empty map on missing/corrupt file.
func LoadCache(path string) map[string][]string {
	out := map[string][]string{}
	data, err := os.ReadFile(path)
	if err != nil {
		return out
	}
	_ = json.Unmarshal(data, &out)
	return out
}

// SaveCache writes the cache atomically (mkdir + sorted JSON).
func SaveCache(path string, cache map[string][]string) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(cache, "", "  ")
	if err != nil {
		return err
	}
	tmp, err := os.CreateTemp(filepath.Dir(path), ".contributors-*.json.tmp")
	if err != nil {
		return err
	}
	tmpName := tmp.Name()
	if _, err := tmp.Write(data); err != nil {
		tmp.Close()
		if rmErr := os.Remove(tmpName); rmErr != nil && !errors.Is(rmErr, os.ErrNotExist) {
			return fmt.Errorf("write temp cache: %w (cleanup temp file: %v)", err, rmErr)
		}
		return err
	}
	if err := tmp.Close(); err != nil {
		if rmErr := os.Remove(tmpName); rmErr != nil && !errors.Is(rmErr, os.ErrNotExist) {
			return fmt.Errorf("close temp cache: %w (cleanup temp file: %v)", err, rmErr)
		}
		return err
	}
	if err := os.Chmod(tmpName, 0o644); err != nil {
		if rmErr := os.Remove(tmpName); rmErr != nil && !errors.Is(rmErr, os.ErrNotExist) {
			return fmt.Errorf("chmod temp cache: %w (cleanup temp file: %v)", err, rmErr)
		}
		return err
	}
	return os.Rename(tmpName, path)
}

// Repo is the minimal interface a caller must provide for enrichment.
type Repo struct {
	Name     string
	PushedAt string
}

// Result is the per-repo enriched output.
type Result struct {
	Repo         Repo
	Contributors []string
}

// FetchFunc retrieves contributor logins via the gh REST endpoint.
type FetchFunc func(ctx context.Context, owner, name string) ([]string, error)

// ShellFetch is the production FetchFunc that shells out to gh api.
func ShellFetch(ctx context.Context, owner, name string) ([]string, error) {
	cmd := exec.CommandContext(ctx, "gh", "api",
		fmt.Sprintf("/repos/%s/%s/contributors?per_page=2", owner, name))
	out, err := cmd.Output()
	if err != nil {
		// 404, 403, timeout etc → empty (parity with Python contributors.py:84).
		var ee *exec.ExitError
		if errors.As(err, &ee) || errors.Is(err, context.DeadlineExceeded) {
			return nil, nil
		}
		return nil, err
	}
	var arr []struct {
		Login string `json:"login"`
	}
	if err := json.Unmarshal(out, &arr); err != nil {
		return nil, nil
	}
	logins := make([]string, 0, len(arr))
	for _, e := range arr {
		if e.Login != "" {
			logins = append(logins, e.Login)
		}
	}
	return StripBots(logins), nil
}

// Enrich fetches contributors for every repo (parallel, bounded).
// Cache hits skip network; misses go through fetch.
// Cache file is rewritten once at the end.
func Enrich(ctx context.Context, owner string, repos []Repo, cachePath string, fetch FetchFunc, maxWorkers int) ([]Result, error) {
	if maxWorkers < 1 {
		maxWorkers = 1
	}
	cache := LoadCache(cachePath)
	results := make([]Result, len(repos))
	var (
		mu       sync.Mutex
		newCache = map[string][]string{}
		wg       sync.WaitGroup
		sem      = make(chan struct{}, maxWorkers)
		firstErr error
	)
	for i, repo := range repos {
		wg.Add(1)
		go func(i int, repo Repo) {
			defer wg.Done()
			select {
			case sem <- struct{}{}:
				defer func() { <-sem }()
			case <-ctx.Done():
				mu.Lock()
				if firstErr == nil {
					firstErr = ctx.Err()
				}
				mu.Unlock()
				return
			}
			key := CacheKey(repo.Name, repo.PushedAt)
			mu.Lock()
			cached, ok := cache[key]
			mu.Unlock()
			var contribs []string
			if ok {
				contribs = append([]string(nil), cached...)
			} else {
				fetched, err := fetch(ctx, owner, repo.Name)
				if err != nil {
					mu.Lock()
					if firstErr == nil {
						firstErr = err
					}
					mu.Unlock()
					return
				}
				contribs = fetched
			}
			mu.Lock()
			newCache[key] = contribs
			results[i] = Result{Repo: repo, Contributors: contribs}
			mu.Unlock()
		}(i, repo)
	}
	wg.Wait()
	if firstErr != nil {
		return nil, firstErr
	}
	// Merge: existing cache entries we hit + new entries.
	for k, v := range newCache {
		cache[k] = v
	}
	if err := SaveCache(cachePath, cache); err != nil {
		return nil, fmt.Errorf("save cache: %w", err)
	}
	return results, nil
}
